File: OBSOLETE/YourBible/plugins/com-shortsands-audioplayer/src/ios/AudioPlayer/AudioBible.swift
//
//  AudioBible.swift
//  AudioPlayer
//
//  Created by Gary Griswold on 7/31/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//

import AVFoundation
#if USE_FRAMEWORK
    import AWS
#endif

class AudioBible {
    
    private static var instance: AudioBible?
    static func shared(controller: AudioBibleController) -> AudioBible {
        if (AudioBible.instance == nil) {
            AudioBible.instance = AudioBible(controller: controller)
        }
        return AudioBible.instance!
    }
    
    private let controller: AudioBibleController
    private let controlCenter: AudioControlCenter
    private var audioAnalytics: AudioAnalytics?
    private var player: AVPlayer?
    // Transient Variables
    private var currReference: AudioReference?
    private var nextReference: AudioReference?
    
    private init(controller: AudioBibleController) {
        self.controller = controller
        self.controlCenter = AudioControlCenter.shared
        self.controlCenter.setupControlCenter(player: self)
        self.initNotifications()
        print("***** Init AudioBible *****")
    }
    
    deinit {
        print("***** Deinit AudioBible *****")
    }
    
    func getPlayer() -> AVPlayer? {
        return self.player
    }
    
    func getCurrentReference() -> AudioReference? {
        return self.currReference
    }
    
    func isPlaying() -> Bool {
        return (self.player != nil) ? self.player!.rate > 0.0 : false
    }
    
    func beginReadFile(reference: AudioReference) {
        print("BibleReader.BEGIN Read File")
        self.currReference = reference
        self.audioAnalytics = AudioAnalytics(mediaSource: reference.tocAudioBook.testament.bible.mediaSource,
                                             mediaId: reference.damId,
                                             languageId: reference.dpbLanguageCode,
                                             textVersion: reference.textVersion,
                                             silLang: reference.silLang)
        self.readVerseMetaData(reference: reference)
        print("INSIDE BibleReader \(reference.damId)")
        AudioPlayState.retrieve(mediaId: reference.damId)
        AwsS3Cache.shared.readFile(s3Bucket: reference.getS3Bucket(),
                   s3Key: reference.getS3Key(),
                   expireInterval: Double.infinity,
                   getComplete: { [unowned self] url in
                    if let audioURL = url {
                        self.initAudio(url: audioURL, reference: reference)
                    }
        })
    }
    
    private func initAudio(url: URL, reference: AudioReference) {
        let asset = AVAsset(url: url)
        let playerItem = AVPlayerItem(asset: asset)
        print("Player Item Status \(playerItem.status)")
        
        let seekTime = backupSeek(state: AudioPlayState.currentState, reference: reference)
        if (CMTimeGetSeconds(seekTime) > 0.1) {
            playerItem.seek(to: seekTime)
        }
        self.player = AVPlayer(playerItem: playerItem)
        self.player?.actionAtItemEnd = AVPlayerActionAtItemEnd.none
        
        self.play()
        self.controller.playHasStarted()
        self.controlCenter.nowPlaying(player: self)
        
        self.preFetchNextChapter(reference: reference)
    }
    
    private func backupSeek(state: AudioPlayState, reference: AudioReference) -> CMTime {
        if (state.mediaUrl == reference.toString()) {
            if reference.audioChapter != nil {
                return state.position
            } else {
                let duration: Int64 = Int64(Date().timeIntervalSince(state.timestamp))
                let backupSec: Int = String(duration).count // could multiply by a factor here
                let backupTime: CMTime = CMTimeMake(Int64(backupSec), 1)
                return(CMTimeSubtract(state.position, backupTime))
            }
        } else {
            return kCMTimeZero
        }
    }
    
    func play() {
        if let play = self.player {
            if let reference = self.currReference {
                print("\n*********** PLAY *************")
                play.play()
                self.audioAnalytics?.playStarted(item: reference.toString(), position: play.currentTime())
            }
        }
    }
    
    func pause() {
        if let play = self.player {
            if let reference = self.currReference {
                print("\n*********** PAUSE *************")
                play.pause()
                self.audioAnalytics?.playEnded(item: reference.toString(), position: play.currentTime())
                self.updateMediaPlayStateTime(reference: reference)
            }
        }
    }
    
    func stop() {
        if self.player != nil && self.player!.rate > 0.0 {
            self.pause()
        }
        self.controller.playHasStopped()
    }
    
    /** This method is called by AudioControlCenter.swift when user clicks the next button. */
    func nextChapter() {
        if let play = self.player {
            let startPlayRate = play.rate
            if let curr = self.nextReference {
                self.currReference = curr
                self.addNextChapter(reference: curr)
                self.preFetchNextChapter(reference: curr)
                if startPlayRate == 0 {
                    self.audioAnalytics?.playStarted(item: curr.toString(), position: play.currentTime())
                }
            } else {
                self.stop()
                AudioPlayState.clear() // Must do after stop, because stop updates
            }
        }
    }
    
    /** This method is called by AudioControlCenter.swift when user clicks the prior button. */
    func priorChapter() {
        if let play = self.player {
            let startPlayRate = play.rate
            if let item = play.currentItem {
                if item.currentTime().seconds < 1.0 {
                    if let prior = self.currReference?.priorChapter() {
                        self.nextReference = self.currReference
                        self.currReference = prior
                        self.addNextChapter(reference: prior)
                        if startPlayRate == 0 {
                            self.audioAnalytics?.playStarted(item: prior.toString(), position: play.currentTime())
                        }
                    }
                } else {
                    item.seek(to: kCMTimeZero)
                    self.controlCenter.nowPlaying(player: self)
                    if startPlayRate == 0 {
                        if let curr = self.currReference {
                            self.audioAnalytics?.playStarted(item: curr.toString(), position: kCMTimeZero)
                        }
                    }
                }
            }
            //play.currentItem?.seek(to: kCMTimeZero)
        }
    }
    
    private func initNotifications() {
        let notify = NotificationCenter.default
        notify.addObserver(self,
                           selector: #selector(playerItemDidPlayToEndTime(note:)),
                           name: .AVPlayerItemDidPlayToEndTime,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(playerItemFailedToPlayToEndTime(note:)),
                           name: .AVPlayerItemFailedToPlayToEndTime,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(playerItemPlaybackStalled(note:)),
                           name: .AVPlayerItemPlaybackStalled,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(playerItemNewErrorLogEntry(note:)),
                           name: .AVPlayerItemNewErrorLogEntry,
                           object: nil)
        /*notify.addObserver(self,
                            selector: #selector(playerItemNewAccessLogEntry(note:)),
                            name: .AVPlayerItemNewAccessLogEntry,
                            object: nil)*/
        /*notify.addObserver(self,
                            selector: #selector(playerItemTimeJumped(note:)),
                            name: .AVPlayerItemTimeJumped,
                            object: nil)*/
        notify.addObserver(self,
                           selector: #selector(applicationDidFinishLaunching(note:)),
                           name: .UIApplicationDidFinishLaunching,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(applicationWillEnterForeground(note:)),
                           name: .UIApplicationWillEnterForeground,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(applicationDidEnterBackground(note:)),
                           name: .UIApplicationDidEnterBackground,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(applicationWillTerminate(note:)),
                           name: .UIApplicationWillTerminate,
                           object: nil)
    }

    @objc private func playerItemDidPlayToEndTime(note:Notification) {
        print("\n** DID PLAY TO END \(String(describing: note.object))")
        self.nextChapter()
    }
    @objc private func playerItemFailedToPlayToEndTime(note:Notification) {
        print("\n********* FAILED TO PLAY TO END *********\(String(describing: note.object))")
    }
    @objc private func playerItemPlaybackStalled(note:Notification) {
        print("\n****** PLAYBACK STALLED \(String(describing: note.object))")
    }
    @objc private func playerItemNewErrorLogEntry(note:Notification) {
        print("\n****** ERROR LOG ENTRY \(String(describing: note.object))\n\(String(describing: self.player?.currentItem?.errorLog()))")
    }
    @objc private func playerItemNewAccessLogEntry(note:Notification) {
        print("\n****** ACCESS LOG ENTRY \(String(describing: note.object))\n\(String(describing: self.player?.currentItem?.accessLog()))")
    }
    @objc private func playerItemTimeJumped(note:Notification) {
        print("\n****** TIME JUMPED \(String(describing: note.object))")
    }
    
    /**
    * This method is called when an App is first, launched, not when it is restarted in foreground
    */
    @objc private func applicationDidFinishLaunching(note:Notification) {
        print("\n****** APP DID FINISH LAUNCHING \(String(describing: note.object)) \(Date().timeIntervalSince1970)")
    }
    /**
    * This method is called when an already running App, which went to background, restarts in foreground.
    */
    @objc private func applicationWillEnterForeground(note:Notification) {
        print("\n****** APP WILL ENTER FOREGROUND \(String(describing: note.object)) \(Date().timeIntervalSince1970)")
    }
    /**
    * This method is called when the Home button or Lock button terminate the foreground App
    */
    @objc private func applicationDidEnterBackground(note:Notification) {
        print("\n****** APP DID ENTER BACKGROUND \(String(describing: note.object)) \(Date().timeIntervalSince1970)")
    }
    /**
    * This method is called when the App is fully killed.  It is not affected by the Home or Lock button
    */
    @objc private func applicationWillTerminate(note:Notification) {
        print("\n****** APP WILL TERMINATE \(String(describing: note.object)) \(Date().timeIntervalSince1970)")
        self.stop()
    }
    
    private func updateMediaPlayStateTime(reference: AudioReference) {
        var result: CMTime = kCMTimeZero
        if let currentTime = self.player?.currentTime() {
            if let audioChapter = reference.audioChapter {
                let verse: Int = audioChapter.findVerseByPosition(priorVerse: 1, time: currentTime)
                result = audioChapter.findPositionOfVerse(verse: verse)
            } else {
                result = currentTime
            }
        }
        AudioPlayState.update(url: reference.toString(), time: result)
    }
    
    private func addNextChapter(reference: AudioReference) {
        self.player?.pause()
        AwsS3Cache.shared.readFile(s3Bucket: reference.getS3Bucket(),
                                   s3Key: reference.getS3Key(),
                                   expireInterval: Double.infinity,
                                   getComplete: { [unowned self] url in
                                    if let audioURL = url {
                                        let asset = AVAsset(url: audioURL)
                                        let playerItem = AVPlayerItem(asset: asset)
                                        self.player?.replaceCurrentItem(with: playerItem)
                                        self.player?.play()
                                        self.readVerseMetaData(reference: reference)
                                        self.controlCenter.nowPlaying(player: self)
                                    }
        })
    }
    
    private func preFetchNextChapter(reference: AudioReference) {
        self.nextReference = reference.nextChapter()
        if let next = self.nextReference {
            AwsS3Cache.shared.readFile(s3Bucket: next.getS3Bucket(),
                                       s3Key: next.getS3Key(),
                                       expireInterval: Double.infinity,
                                       getComplete: { url in })
        }
    }
    
    private func readVerseMetaData(reference: AudioReference) {
        reference.audioChapter = nil
        if let reader = self.controller.metaDataReader {
            reader.readVerseAudio(damid: reference.damId, bookId: reference.bookId, chapter: reference.chapterNum,
                                  complete: { audioChapter in
                reference.audioChapter = audioChapter
            })
        }
    }
}

File: OBSOLETE/YourBible/plugins/com-shortsands-audioplayer/src/android/app/src/main/java/com/shortsands/audioplayer/AudioBible.java
package com.shortsands.audioplayer;

import android.media.AudioManager;
import android.media.MediaPlayer;
import android.os.PowerManager;
import android.util.Log;
import com.shortsands.aws.AwsS3Cache;
import com.shortsands.aws.CompletionHandler;
import java.io.File;
import java.io.IOException;

/**
 * Created by garygriswold on 8/30/17.
 */

public class AudioBible implements MediaPlayer.OnErrorListener, MediaPlayer.OnCompletionListener,
        MediaPlayer.OnSeekCompleteListener {

    private static final String TAG = "AudioBible";

    private static AudioBible instance = null;
    static AudioBible shared(AudioBibleController controller) {
        if (AudioBible.instance == null) {
            AudioBible.instance = new AudioBible(controller);
        }
        return(AudioBible.instance);
    }

    private final AudioBibleController controller;
    private final AudioControlCenter controlCenter;
    private AudioAnalytics audioAnalytics;
    // Transient Variables
    private AudioReference currReference;
    private AudioReference nextReference;
    private MediaPlayer mediaPlayer;
    private MediaPlayer nextPlayer;

    private AudioBible(AudioBibleController controller) {
        super();
        this.controller = controller;
        this.controlCenter = AudioControlCenter.shared;
    }

    MediaPlayer getPlayer() {
        return this.mediaPlayer;
    }

    AudioReference getCurrReference() {
        return this.currReference;
    }

    boolean isPlaying() {
        return (this.mediaPlayer != null) ? this.mediaPlayer.isPlaying() : false;
    }

    void beginReadFile(AudioReference reference) {
        Log.d(TAG, "BibleReader.BEGIN Read File");
        this.currReference = reference;
        this.audioAnalytics = new AudioAnalytics(controller.activity,
            reference.tocAudioBook.testament.bible.mediaSource,
            reference.damId(),
            reference.dbpLanguageCode(),
            reference.textVersion(),
            reference.silLang());
        BeginReadFileCompletion handler = new BeginReadFileCompletion();
        this.readVerseMetaData(reference);
        AudioPlayState.retrieve(this.controller.activity, reference.damId());
        AwsS3Cache.shared().readFile(reference.getS3Bucket(),
                reference.getS3Key(),
                Integer.MAX_VALUE,
                handler);
    }

    class BeginReadFileCompletion implements CompletionHandler {
        @Override
        public void completed(Object result) {
            if (result instanceof File) {
                File file = (File) result;
                initAudio(file);
            }
        }
        @Override
        public void failed(Throwable exception) {
            Log.d(TAG, "BeginReadFile Failed " + exception.toString());
        }
    }

    private void initAudio(File url) {
        if (url != null) {
            this.mediaPlayer = this.initPlayer(url);
            if (this.mediaPlayer != null) {
                long seekTime = this.backupSeek(AudioPlayState.currentState, this.currReference);
                if (seekTime > 100L) {
                    this.mediaPlayer.setOnSeekCompleteListener(this);
                    this.mediaPlayer.seekTo((int) seekTime);
                } else {
                    this.onSeekComplete(this.mediaPlayer);
                }
            }
        } else {
            Log.e(TAG, "URL is null");
        }
    }

    private long backupSeek(AudioPlayState state, AudioReference reference) {
        if (state.mediaUrl != null && state.mediaUrl.equals(reference.toString())) {
            if (reference.audioChapter != null) {
                return state.position;
            } else {
                long duration = System.currentTimeMillis() - state.timestamp;
                long backupSec = String.valueOf(duration / 1000L).length() * 1000L;  // could multiply by a factor here
                return(state.position - backupSec);
            }
        } else {
            return 0L;
        }
    }

    @Override
    public void onSeekComplete(MediaPlayer player) {
        player.setOnSeekCompleteListener(null);
        this.play();
        player.setOnCompletionListener(this);
        this.controller.playHasStarted(player);
        this.controlCenter.nowPlaying(this);
        this.preFetchNextChapter(this.currReference);
    }

    private MediaPlayer initPlayer(File file) {
        String url = file.getAbsolutePath();
        MediaPlayer player = new MediaPlayer();
        player.setLooping(false);
        player.setAudioStreamType(AudioManager.STREAM_VOICE_CALL); // Replace with AudioAttributes in SDK 21
        try {
            player.setDataSource(url);
            player.prepare();
            player.setWakeMode(this.controller.activity, PowerManager.PARTIAL_WAKE_LOCK);
            return player;
        } catch(IOException ioe) {
            Log.e(TAG, "Error in AudioBible.initPlayer " + ioe.toString());
            return null;
        }
    }

    void play() {
        Log.d(TAG, "\n*********** PLAY *************");
        if (this.mediaPlayer != null) {
            this.mediaPlayer.start();
            long currTime = this.mediaPlayer.getCurrentPosition();
            this.audioAnalytics.playStarted(this.currReference.toString(), currTime);
        }
    }

    void pause() {
        Log.d(TAG, "\n*********** PAUSE *************");
        if (this.mediaPlayer != null) {
            this.mediaPlayer.pause();
            long currTime = this.mediaPlayer.getCurrentPosition();
            this.audioAnalytics.playEnded(this.currReference.toString(), currTime);
            this.updateMediaPlayStateTime(this.currReference);
        }
    }

    void stop() {
        if (this.mediaPlayer != null && this.mediaPlayer.isPlaying()) {
            this.pause();
        }
        this.controller.playHasStopped();
        if (this.mediaPlayer != null) {
            this.mediaPlayer.reset();
            this.mediaPlayer.release();
            this.mediaPlayer = null;
        }
        if (this.nextPlayer != null) {
            this.nextPlayer.reset();
            this.nextPlayer.release();
            this.nextPlayer = null;
        }
    }

    /** This method is called by AudioControlCenter, which is not implemented in the android version */
    void nextChapter() {
        if (this.nextReference != null) {
            this.currReference = this.nextReference;
            this.addNextChapter(this.currReference);
            this.preFetchNextChapter(this.currReference);
        } else {
            this.stop();
            AudioPlayState.clear(this.controller.activity); // Must be after stop, because stop does update
        }
    }

    /** This method is called by AudioControlCenter, which is not implemented in the android version */
    void priorChapter() {

    }

    @Override
    public void onCompletion(MediaPlayer player) {
        player.setOnCompletionListener(null);
        this.nextChapter();
    }

    @Override
    public boolean onError(MediaPlayer player, int what, int extra) {
        //this.errorTime = System.currentTimeMillis();
        player.setOnErrorListener(null);
        String message;
        switch (what) {
            case MediaPlayer.MEDIA_ERROR_IO:
                message = "MEDIA ERROR IO";
                break;
            case MediaPlayer.MEDIA_ERROR_MALFORMED:
                message = "MEDIA ERROR MALFORMED";
                break;
            case MediaPlayer.MEDIA_ERROR_NOT_VALID_FOR_PROGRESSIVE_PLAYBACK:
                message = "MEDIA ERROR NOT VALID FOR PROGRESSIVE PLAYBACK";
                break;
            case MediaPlayer.MEDIA_ERROR_SERVER_DIED:
                message = "MEDIA ERROR SERVER DIED";
                break;
            case MediaPlayer.MEDIA_ERROR_TIMED_OUT:
                message = "MEDIA ERROR TIMED OUT";
                break;
            case MediaPlayer.MEDIA_ERROR_UNKNOWN:
                message = "MEDIA ERROR UNKNOWN";
                break;
            case MediaPlayer.MEDIA_ERROR_UNSUPPORTED:
                message = "MEDIA ERROR UNSUPPORTED";
                break;
            default:
                message = "Unknown Error " + what;
        }
        Log.e(TAG, "onError " + message + " " + extra);

        player.reset();
        player.setOnErrorListener(this);
        //this.onErrorRecovery = true;
        //this.progressBar.setVisibility(View.VISIBLE);
        //Uri videoUri = Uri.parse(this.videoUrl);
        //this.videoView.setVideoURI(videoUri);
        return true;
    }

    private void updateMediaPlayStateTime(AudioReference reference) {
        AudioTOCChapter chapter = reference.audioChapter;
        int position = this.mediaPlayer.getCurrentPosition();
        if (chapter != null) {
            int verseNum = chapter.findVerseByPosition(1, position);
            position = chapter.findPositionOfVerse(verseNum);
        } else {
            position = (position >= 0) ? position : 0;
        }
        AudioPlayState.update(this.controller.activity, reference.toString(), position);
    }

    private void addNextChapter(AudioReference reference) {
        if (this.mediaPlayer.isPlaying()) {
            this.mediaPlayer.stop();
            this.mediaPlayer = this.nextPlayer;
            this.mediaPlayer.start();
        } else {
            this.mediaPlayer = this.nextPlayer;
        }
        this.controller.nextMediaPlayer(this.mediaPlayer);
        this.mediaPlayer.setOnCompletionListener(this);
        this.readVerseMetaData(reference);
        this.controlCenter.nowPlaying(this);
    }

    private void preFetchNextChapter(AudioReference reference) {
        this.nextReference = reference.nextChapter();
        if (this.nextReference != null) {
            PreFetchCompletion handler = new PreFetchCompletion();
            AwsS3Cache.shared().readFile(this.nextReference.getS3Bucket(),
                    this.nextReference.getS3Key(),
                    Integer.MAX_VALUE,
                    handler);
        } else {
            nextPlayer = null;
        }
    }
    class PreFetchCompletion implements CompletionHandler {
        @Override
        public void completed(Object result) {
            if (result instanceof File) {
                File file = (File) result;
                nextPlayer = initPlayer(file);
                mediaPlayer.setNextMediaPlayer(nextPlayer);
            }
        }
        @Override
        public void failed(Throwable exception) {
            Log.d(TAG, "NextReadFile Failed " + exception.toString());
        }
    }

    private void readVerseMetaData(AudioReference reference) {
        AudioTOCBible reader = reference.tocAudioBook.testament.bible;
        AudioTOCChapter chapter = reader.readVerseAudio(reference.damId(), reference.bookId(), reference.chapterNum());
        reference.audioChapter = chapter;
        if (chapter == null) {
           Log.d(TAG, "Unable to read verse position data");
        }
    }
}
